src/libstd/collections/hash/map.rs

```diff
@@ -1208,6 +1208,34 @@ impl<K, V, S> HashMap<K, V, S>
 
         self.search_mut(k).into_occupied_bucket().map(|bucket| pop_internal(bucket).1)
     }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// In other words, remove all pairs `(k, v)` such that `f(&k,&v)` returns `false`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::HashMap;
+    ///
+    /// let xs = [(1, 10), (2, 20), (3, 30), (4, 40), (5, 50), (6, 60)];
+    /// let mut map: HashMap<isize, isize> = xs.iter().cloned().collect();
+    /// map.retain(|&k, _| k % 2 == 0);
+    /// assert_eq!(map.len(), 3);
+    /// ```
+    #[stable(feature = "retain_hash_map", since = "1.17.0")]
```

Comment:
What we learned from `Vec::retain`'s restrictions is that mutable access while retaining is useful. In this case it should then be that it passes `&K, &mut V`.